### DISCUSSION TITLE
Wrong types with mobx `values()` result in Getting started tutorial

### DISCUSSION THREAD

**Discussion by alexamy** [mobxjs/mobx-state-tree#1981] [Q&A] [answered]
Hi! I'm trying to complete the [Getting started](https://mobx-state-tree.js.org/intro/getting-started) tutorial, but got typing error from mobx `values`.

At the step [Getting to the UI](https://mobx-state-tree.js.org/intro/getting-started#getting-to-the-ui) we are using `values` from mobx to iterate through map from store:
```typescript
// in react component
{values(props.store.todos).map(todo => (
    <div>
        <input type="checkbox" checked={todo.done} onChange={e => todo.toggle()} />
        <input type="text" value={todo.name} onChange={e => todo.setName(e.target.value)} />
    </div>
))}
```
Store:
```typescript
// somewhere in store
const Todo = types.model({
  name: '',
  done: false,
})

const Store = types.model({
  todos: types.map(Todo),
});
```

This code works in the sandbox, but I got the typing error about `name` property (and possible undefined `todo`):
![image](https://user-images.githubusercontent.com/596195/201020601-2c12fd4c-0bca-4ca5-baa3-3f9ac2182a26.png)

I've searched and found an old [issue](https://github.com/mobxjs/mobx-state-tree/issues/783), but there is no typings discussion.
Official playground examples use JS.

_What am I doing wrong?_

[Playground](https://codesandbox.io/s/mobx-state-tree-playground-broken-values-7s9ujq?file=/src/Todos.tsx) with code from tutorial
[Playground](https://codesandbox.io/s/mobx-state-tree-playground-broken-values-simplified-dvzo4o?file=/src/App.tsx) with simplified version (in one file)

**Comment by EmilTholin** [comment, marked as answer]
Hi @alexamy!

You can use the `values` method on your map to get an iterator object and then create an array from that to retain the type information:

```ts
export default function App() {
  return (
    <div className='App'>
      {Array.from(store.todos.values()).map((todo, i) => (
        <p key={i}>{todo.name}</p>
      ))}
    </div>
  );
}
```

Nice find though, we should consider updating the documentation to retain the type information, even though the example is in plain JavaScript.

**Reply by alexamy** [reply]
So all [utilities](https://mobx.js.org/api.html#collection-utilities-) from mobx potentially can break typings, and we should not use them? 
It looks like MST functionality (actions, etc) and native js methods can cover all use cases of these utilites.

**Reply by alexamy** [reply]
Btw quote from getting started:
> MST loves MobX, and is fully compatible with it's autorun, reaction, observe and other parts of the API.

**Reply by EmilTholin** [reply]
I have personally not used the MobX `values` function when working with MST, since maps have that method itself, so I didn't know the type information was not retained.

I agree with the quote though; `autorun`, `reaction`, `observe` still works nicely with MST, even though sometimes pure MST concepts might be a better fit, but not always.
